### PR TITLE
Ignore gossip AggregateAndProof if aggregate is seen

### LINKED ISSUE #3794: Ignore AggregateAndProof with duplicate aggregated attestations
**Is your feature request related to a problem? Please describe.**

For same aggregated attestations, there are up to 16 different gossip messages with the aggregated attestations.

**Describe the solution you'd like**

+ Ignore if we've already seen the aggregated attestations from a gossip AggregateAndProof message

+ Make sure we have better metrics for AggregateAndProof gossip topics

https://github.com/ethereum/consensus-specs/pull/2838

### PR DESCRIPTION
**Pull request by twoeths**
**Motivation**

Address this spec when validating `beacon_aggregate_and_proof` gossip message:
```
[IGNORE] The valid aggregate attestation defined by hash_tree_root(aggregate) has not already been seen (via aggregate gossip, within a verified block, or through the creation of an equivalent aggregate locally).
```

**Description**

+ Implement `SeenAggregatedAttestations`
+ Populate seen aggregated attestations from api / gossip / blocks
+ Check if the aggregated attestation is seen when validating gossip `AggregateAndProof` message

Closes #3794

**TODO**
+ [x] Deploy and test in a node, make sure it takes less time to validate AggregateAndProof gossip messages

### CODE REVIEW

**Comment by dapplion** [comment]
This PR is a subset of https://github.com/ethereum/consensus-specs/pull/2847 right?

This PR ignores only exact "copies" of received aggregates. Not sure if the plan is to keep both conditions

**Comment by twoeths** [comment]
> This PR is a subset of https://github.com/ethereum/consensus-specs/pull/2847 right?

yeah it's just implemented what's already merged in the spec v1.1.10, thanks for pointing me to the PR, I can make it one step further to ignore attestations with subset aggregation bits. The performance should be even better!

**Comment by twoeths** [comment]
This PR shows a better performance of lodestar overall

- Memory is saved 10%
Screen Shot 2022-05-17 at 09 22 30

- CPU is saved 10% - 20%
Screen Shot 2022-05-17 at 09 23 23

- Less signature verification
Screen Shot 2022-05-17 at 09 24 54

- Better performance of AggregateAndProof gossip validation
Screen Shot 2022-05-17 at 09 25 39

- Gossip block processed time is improved too
Screen Shot 2022-05-17 at 09 26 56

- 75% of AggregateAndProof gossip messages were ignored
Screen Shot 2022-05-17 at 09 28 02

- Gossipsub-js forwarded less AggregateAndProof messages
Screen Shot 2022-05-17 at 09 29 36

**Comment by dapplion** [comment]
@tuyennhv There's are some important nuances in how to consider an aggregate overlapping

From https://github.com/ethereum/consensus-specs/pull/2847 

> it does not mandate dropping aggregates where a union of previous aggregates would cause it to become a subset

Do we have to check against the union of bitfields from all aggregates, or each bitfield from each aggregate? Raised this question in https://github.com/ethereum/consensus-specs/pull/2847#issuecomment-1129567258

**Comment by dapplion** [comment]
This PR can re-use the BitArray set comparison logic from https://github.com/ChainSafe/lodestar/pull/4034

**Comment by twoeths** [comment]
Metrics show that it's less than 2 `isNonStrictSuperSet` calls per `isKnown` since we sorted by `trueBitCount` desc

Screen Shot 2022-05-20 at 11 15 39

Perf test shows that the "isKnown" api is cheap
```
SeenAggregatedAttestations perf test
    ✓ isKnown best case - 1 super set check                                2398082 ops/s    417.0000 ns/op        -     291100 runs   1.31 s
    ✓ isKnown normal case - 2 super set checks                             2487562 ops/s    402.0000 ns/op        -      89806 runs  0.404 s
    ✓ isKnown worse case - 16 super set checks                             2493766 ops/s    401.0000 ns/op        -     111188 runs  0.506 s
```

File: packages/lodestar/src/chain/seenCache/seenAggregateAndProof.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  private readonly aggregateRootsByEpoch = new MapDef<Epoch, MapDef<RootHex, AggregationInfo[]>>(
+    () => new MapDef<RootHex, AggregationInfo[]>(() => [])
+  );
+  private lowestPermissibleEpoch: Epoch = 0;
+
+  constructor(private readonly metrics: IMetrics | null) {}
+
+  isKnown(targetEpoch: Epoch, attDataRoot: RootHex, aggregationBits: BitArray): boolean {
+    const seenAggregationInfoArr = this.aggregateRootsByEpoch.getOrDefault(targetEpoch).getOrDefault(attDataRoot);
+    this.metrics?.seenCache.aggregatedAttestations.isKnownTotal.inc();
```
Comment (dapplion) [inline packages/lodestar/src/chain/seenCache/seenAggregateAndProof.ts]:
```suggestion
    this.metrics?.seenCache.aggregatedAttestations.isKnownCalls.inc();
```

File: packages/lodestar/src/chain/seenCache/seenAggregateAndProof.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    for (let i = 0; i < seenAggregationInfoArr.length; i++) {
+      if (isNonStrictSuperSet(seenAggregationInfoArr[i].aggregationBits, aggregationBits)) {
+        this.metrics?.seenCache.aggregatedAttestations.superSetCheckTotal.observe(i + 1);
+        this.metrics?.seenCache.aggregatedAttestations.isKnownSuccess.inc();
```
Comment (dapplion) [inline packages/lodestar/src/chain/seenCache/seenAggregateAndProof.ts]:
```suggestion
        this.metrics?.seenCache.aggregatedAttestations.isKnownHits.inc();
```

Comment (dapplion) [reply packages/lodestar/src/chain/seenCache/seenAggregateAndProof.ts]: Same for contribution and proofs

**Review by dapplion — APPROVED** [review]
Looks great!

**Pull request merged**